CONTRIBUTING.md

```diff
@@ -456,6 +456,172 @@ checks if contributed code does not have the appropriate license headers.
 > be automatically configured to add the correct license header to new source
 > files based on the source location.
 
+### Logging
+
+The Elasticsearch server logs are vitally useful for diagnosing problems in a
+running cluster. You should make sure that your contribution uses logging
+appropriately: log enough detail to inform users about key events and help them
+understand what happened when things go wrong, without logging so much detail
+that the logs fill up with noise and the useful signal is lost.
+
+Elasticsearch uses Log4J for logging. In most cases you should log via a
+`Logger` named after the class that is writing the log messages, which you can
+do by declaring a static field of the class as follows:
+
+    private static final Logger logger = LogManager.getLogger(MyClassName.class);
+
+There are a number of other acceptable ways to obtain a `Logger`. For example,
+if there will be many instances of a long-lived component whose logs should be
+distinguished then each instance can use a prefixed logger obtained from one of
+the methods in `org.elasticsearch.common.logging.Loggers`.
```

Comment:
An example might be useful here, especially considering 1) this document is going to be consumed by folks who are just starting and not yet familiar with our patterns, and 2) we don't actually use prefixed logger that much compared to regular loggers.

Reply:
The lack of an example was deliberate here. Almost everyone will do the thing I already suggested. Maybe it'd be better to have even fewer words here and just link to the `Loggers` class. You can always navigate to the class and see how/where it's used if you're interested.